pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,483 @@ func (c *Controller) processNextWorkItem() bool {
-		tenant.Status.Result = api.TenantResultSuccess
-		tenant.Status.Message = "Tenant namespace successfully prepared"
-		if tenant, err = c.updateStatus(tenant); err != nil {
-			return err
-		}
-		tenant, _ = c.updateProgress(tenant, api.TenantProgressFinished)
-		log.Printf("Tenant preparation successful for %s", tenant.GetName())
+func (c *Controller) reconcile(config clientConfig, tenant *api.Tenant) (err error) {
+	if c.isInitialized(tenant) {
+		err = c.reconcileInitialized(config, tenant)
+	} else {
+		err = c.reconcileUninitialized(config, tenant)
 	}
-	c.updateMetrics()
-	c.syncCount++
+	return
+}
+
+func (c *Controller) reconcileUninitialized(config clientConfig, tenant *api.Tenant) error {
+	c.logPrintln(tenant, "tenant not initialized yet")
+
+	nsName, err := c.createTenantNamespace(config, tenant)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
+		tenant.Status.SetCondition(&knativeapis.Condition{
+			Type:    knativeapis.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  api.StatusReasonFailed,
+			Message: condMsg,
+		})
+		return err
+	}
+
+	_, err = c.reconcileTenantRoleBinding(tenant, nsName, config)
+	if err != nil {
+		condMsg := fmt.Sprintf("Failed to create the tenant namespace.")
+		tenant.Status.SetCondition(&knativeapis.Condition{
+			Type:    knativeapis.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  api.StatusReasonFailed,
+			Message: condMsg,
+		})
+		c.deleteTenantNamespace(nsName, tenant, config) // clean-up ignoring error
+		return err
+	}
+
+	tenant.Status.TenantNamespaceName = nsName
+
+	tenant.Status.SetCondition(&knativeapis.Condition{
+		Type:   knativeapis.ConditionReady,
+		Status: corev1.ConditionTrue,
+	})
+
 	return nil
 }
 
-func (c *Controller) removeFinalizer(tenant *api.Tenant) error {
-	changed, finalizerList := utils.RemoveString(tenant.ObjectMeta.Finalizers, k8s.FinalizerName)
-	if changed {
-		tenant.ObjectMeta.Finalizers = finalizerList
-		_, err := c.update(tenant)
-		if err != nil {
-			return err
+func (c *Controller) reconcileInitialized(config clientConfig, tenant *api.Tenant) error {
+	c.logPrintln(tenant, "tenant is initialized already")
+
+	nsName := tenant.Status.TenantNamespaceName
+
+	exists, err := c.checkNamespaceExists(nsName)
+	if err != nil {
+		c.logPrintln(tenant, err)
+		return err
+	}
+
+	if !exists {
+		condMsg := fmt.Sprintf(
+			"The tenant namespace %q does not exist anymore."+
+				" This issue must be analyzed and fixed by an operator.",
+			nsName,
+		)
+		tenant.Status.SetCondition(&knativeapis.Condition{
+			Type:    knativeapis.ConditionReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  api.StatusReasonDependentResourceState,
+			Message: condMsg,
+		})
+		err = errors.Errorf("tenant namespace %q does not exist anymore", nsName)
+		c.logPrintln(tenant, err)
+		return err
+	}
+
+	updateNeeded, err := c.reconcileTenantRoleBinding(tenant, nsName, config)
```

Comment:
```suggestion
	changed, err := c.reconcileTenantRoleBinding(tenant, nsName, config)
```

Reply:
See above. `changed` would be wrong, as in case of an error nothing has changed but still an update is needed.